src/evaluators/ForInStatement.js

```diff
@@ -160,30 +181,33 @@ function emitResidualLoopIfSafe(
         }
       });
       if (targetObject instanceof ObjectValue && sourceObject !== undefined) {
-        let oe = ob.values.getElements();
-        if (oe.size !== 1) ob.throwIfNotConcreteObject();
         let o;
-        for (let co of oe) o = co;
-        invariant(o !== undefined);
+        let oe = ob.values.getElements();
+        if (oe.size !== 1) {
+          o = ob;
+        } else {
+          for (let co of oe) o = co;
+          invariant(o !== undefined && o.isSimple());
+        }
         let generator = realm.generator;
         invariant(generator !== undefined);
         // make target object simple and partial, so that it returns a fully
         // abstract value for every property it is queried for.
         targetObject.makeSimple();
         targetObject.makePartial();
-        if (sourceObject === o && sourceObject.isSimple()) {
+        if (sourceObject === o) {
           // Known enumerable properties of sourceObject can become known
           // properties of targetObject.
-          let sourceIsPartial = sourceObject.isPartial();
+          invariant(sourceObject.isPartial());
           sourceObject.makeNotPartial();
           // EnumerableOwnProperties is sufficient because sourceObject is simple
           let keyValPairs = EnumerableOwnProperties(realm, sourceObject, "key+value");
-          if (sourceIsPartial) sourceObject.makePartial();
```

Comment:
Currently it is impossible to write a loop where the source is not partial and that does not generate an error before getting to this point.